src/codecs/framers/octet_counting.rs

```diff
@@ -0,0 +1,394 @@
+use crate::codecs::{BoxedFramer, BoxedFramingError, FramingConfig};
+use bytes::{Buf, Bytes, BytesMut};
+use serde::{Deserialize, Serialize};
+use std::io;
+use tokio_util::codec::{LinesCodec, LinesCodecError};
+
+/// Config used to build a `OctetCountingCodec`.
+#[derive(Debug, Clone, Default, Deserialize, Serialize)]
+pub struct OctetCountingDecoderConfig {
+    max_length: Option<usize>,
+}
+
+#[typetag::serde(name = "octet_counting")]
+impl FramingConfig for OctetCountingDecoderConfig {
+    fn build(&self) -> crate::Result<BoxedFramer> {
+        Ok(Box::new(match self.max_length {
+            Some(max_length) => OctetCountingCodec::new_with_max_length(max_length),
+            None => OctetCountingCodec::new(),
+        }))
+    }
+}
+
+/// Codec using the `Octet Counting` format as specified in
+/// https://tools.ietf.org/html/rfc6587#section-3.4.1.
+#[derive(Clone, Debug)]
+pub struct OctetCountingCodec {
```

Comment:
Do we want to call this a "codec" when it just does framing? And above we use "decoder" in the config struct. I don't have a strong opinion on the naming, but I do think it'd be valuable to be precise and consistent with what we're calling each of these things.

Reply:
It is a "codec" by the tokio terminology. The above "decoder" refers to the decoding side, which we only use in sources. Once we move to the sinks, that codec will also implement an "encoder", and have an additional `OctetCountingEncodingConfig`, and implement `tokio_util::codec::Encode` on `OctetCountingCodec` alongside the existing `tokio_util::codec::Decode`.

We can revisit the whole codec story afterwards with an PR to change the naming everywhere if you feel like that should still change given the previous argument. Currently all framers are called `*Codec` for that reason and have a `*DecodingConfig`, and will add a `*EncodingConfig` in the future. If we call it `*Framer`, not sure how the `*EncodingConfig`/`*DecodingConfig` would be sensibly called.

Reply:
Gotcha, as long as they're well-defined and consistent I'm good with that.